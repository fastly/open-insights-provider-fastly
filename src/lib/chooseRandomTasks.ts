import { TaskData } from "../@types";

function mean(values: number[]): number {
  return values.reduce((memo: number, i): number => memo + i) / values.length;
}

/**
 * This defines a random number between 0 and "mass", which should cover the entire
 * domain of the probability distribution. Then we will "walk" the probability mass
 * until we exceed the generated value
 */
function basicChooseOne(tasks: TaskData[], mass: number): number {
  const randomMass = mass * Math.random();
  let index = 0;
  let weight = tasks[0].weight;
  while (weight < randomMass) {
    index += 1;
    weight += tasks[index].weight;
  }
  return index;
}

function chooseRandomTasks(tasks: TaskData[], k: number): TaskData[] {
  //  If we have more work slots than candidates in the probability distribution,
  //  we return them all
  if (k >= tasks.length) {
    return tasks;
  }

  // Since we are mutating the probabilities vector, we copy it first
  const t = tasks.map((t): TaskData => Object.assign({}, t));

  // This assumes that the probability vector is normalized. If it is not, it
  // should be changed to the sum of the weights
  let remaining = tasks
    .map((t): number => t.weight)
    .reduce((memo, task): number => memo + task);

  // The weight of an additional "ghost" element that we will use to
  // inhibit running tasks that should have low probability
  const probs = tasks.map((t): number => t.weight);
  const dampeningFactor = Math.min(mean(probs), 0.05 * remaining);
  const choices = [];

  for (let j = 0; j < k; j++) {
    // Perform choice adding an additional "ghost" element that can be tweaked
    // to control execution eagerness
    const d = ({ weight: dampeningFactor } as any) as TaskData;
    const i = basicChooseOne(t.concat([d]), remaining + dampeningFactor);

    // If we choose a _real_ task, instead of the _ghost_ task, execute the logic
    // inside the conditional
    if (i < t.length) {
      // Add selected element to the selection set
      choices.push(Object.assign({}, t[i]));

      // Eliminate chosen element from distribution and renormalize
      remaining -= t[i].weight;
      t[i].weight = 0;
    }
  }

  return choices;
}

export default chooseRandomTasks;
