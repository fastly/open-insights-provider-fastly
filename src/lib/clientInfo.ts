import { ClientInfo } from "../@types";

interface Cache {
  [key: string]: Promise<ClientInfo>;
}

const cache: Cache = {};

/** Cleans the cache: necessary for testing */
function reset(url: string): void {
  delete cache[url];
}

// TODO: Need to handle errors and retry
/**
 * Function that performs a unique API query to get client data
 * Memoizes responses to ensure that we only hit the API once
 */
function getClientInfo(url: string): Promise<ClientInfo> {
  if (cache[url]) {
    return cache[url];
  }
  cache[url] = fetch(url).then(
    (r): ClientInfo => (r.json() as unknown) as ClientInfo
  );
  return cache[url];
}

export { getClientInfo, reset };
