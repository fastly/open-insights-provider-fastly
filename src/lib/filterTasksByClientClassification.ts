import { Client, TaskData } from "../@types";

export default function filterTasksByClientClassification(
  tasks: TaskData[],
  client: Client
): TaskData[] {
  return tasks.filter(({ classification }): boolean => {
    return Object.keys(classification).reduce<boolean>((acc, key): boolean => {
      const classifiers = classification[key];
      if (classifiers) {
        return (
          acc &&
          classifiers.some(
            (val: string | number): boolean => val === client[key]
          )
        );
      }
      return acc;
    }, true);
  });
}
