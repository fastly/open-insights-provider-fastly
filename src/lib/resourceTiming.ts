/*eslint guard-for-in:0*/
import { ResourceTimingEntry } from "@openinsights/openinsights";
import compose from "../util/compose";
import camelCaseToSnakeCase from "../util/camelCaseToSnakeCase";

const EXCLUDED_PROPS = ["name", "initiatorType", "entryType"];

function cloneEntry(entry: ResourceTimingEntry): ResourceTimingEntry {
  const result: ResourceTimingEntry = {};
  for (const key in entry) {
    const type = typeof entry[key];
    if (type === "number" || type === "string") {
      result[key] = entry[key];
    }
  }
  return result;
}

function removeEntryProps(
  entry: ResourceTimingEntry,
  props: string[]
): ResourceTimingEntry {
  const result: ResourceTimingEntry = {};
  return Object.keys(entry).reduce((res, key): ResourceTimingEntry => {
    if (props.indexOf(key) < 0) {
      res[key] = entry[key];
    }
    return res;
  }, result);
}

function normalizeEntryKeys(entry: ResourceTimingEntry): ResourceTimingEntry {
  const result: ResourceTimingEntry = {};
  return Object.keys(entry).reduce((res, key): ResourceTimingEntry => {
    const newKey = camelCaseToSnakeCase(key);
    res[newKey] = entry[key];
    return res;
  }, result);
}

function normalizeEntryProps(
  props: string[]
): (entry: ResourceTimingEntry) => ResourceTimingEntry {
  return (entry): ResourceTimingEntry => removeEntryProps(entry, props);
}

const normalizeEntry: (
  entry: ResourceTimingEntry
) => ResourceTimingEntry = compose(
  normalizeEntryKeys,
  normalizeEntryProps(EXCLUDED_PROPS),
  cloneEntry
);

export {
  cloneEntry,
  removeEntryProps,
  normalizeEntryKeys,
  normalizeEntryProps,
  normalizeEntry,
};
