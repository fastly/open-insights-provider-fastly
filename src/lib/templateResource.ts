import { Config } from "../@types";

type ResourceReducer =
  | ((resource: string, config: Config) => string)
  | ((resource: string) => string);

const DICTIONARY =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
const DICTIONARY_LENGTH = DICTIONARY.length;
const RANDOM_LENGTH = 16;

// Removes all spaces from string, this allows the other reducers to have more
// reliable match patterns as they don't need to be concerned about whitespace.
function removeWhitespace(resource: string): string {
  return resource.replace(/\s/g, "");
}

// Replaces the template placeholder <% RANDOM %> with a random string
function replaceRandom(resource: string): string {
  const randomLetter = (): string =>
    DICTIONARY.charAt(Math.floor(Math.random() * DICTIONARY_LENGTH));
  const random = Array(RANDOM_LENGTH)
    .fill(null)
    .reduce((acc): string => acc + randomLetter(), "");
  return resource.replace(/<%RANDOM%>/gi, random);
}

// Replaces the template placeholder <% TEST_ID %> with the test ID from config
function replaceTestId(resource: string, config: Config): string {
  return resource.replace(/<%TEST_ID%>/gi, config.test.id);
}

export function templateResource(resource: string, config: Config): string {
  // Note: removeWhitespace must come first to ensure the proceeding reducers
  // are consitent and deterministic.
  const reducers: ResourceReducer[] = [
    removeWhitespace,
    replaceTestId,
    replaceRandom,
  ];
  return reducers.reduce((acc, r): string => r(acc, config), resource);
}
