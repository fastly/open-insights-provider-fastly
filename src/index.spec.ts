import fetchMock from "jest-fetch-mock";
import { mockRandomForEach } from "jest-mock-random";

import { Fetch, TestResultBundle } from "@openinsights/openinsights";
import { Provider } from "./";
import * as template from "./lib/templateResource";

import configFixture from "./fixtures/config";
import clientInfoFixture from "./fixtures/clientInfo";
import entryFixture from "./fixtures/entry";

const settingsFixture = configFixture.settings;
const taskFixture = configFixture.tasks[1];

describe("FastlyProvider", (): void => {
  let providerFixture: Provider;

  beforeEach((): void => {
    providerFixture = new Provider(settingsFixture) as Provider;
    providerFixture.setSessionConfig(configFixture);
    fetchMock.resetMocks();
  });

  describe("shouldRun", (): void => {
    it("should always run", () => {
      const result = new Provider(settingsFixture);
      expect(result.shouldRun()).toBeTruthy();
    });
  });

  describe("fetchSessionConfig", (): void => {
    it("should fetch session config", (): Promise<void> => {
      fetchMock.mockResponseOnce(JSON.stringify(configFixture), {
        headers: {
          "access-control-allow-origin": "*",
        },
      });
      const provider = new Provider(settingsFixture);
      return provider.fetchSessionConfig().then((result) => {
        expect(result).toEqual(configFixture);
      });
    });

    it("should retry the fetch if it fails", (): Promise<void> => {
      fetchMock.mockAbortOnce();
      fetchMock.mockResponseOnce(JSON.stringify(configFixture), {
        headers: {
          "access-control-allow-origin": "*",
        },
      });
      const provider = new Provider(settingsFixture);
      return provider.fetchSessionConfig().then((result) => {
        expect(fetchMock.mock.calls).toHaveLength(2);
        expect(result).toEqual(configFixture);
      });
    });
  });

  describe("expandTasks", (): void => {
    mockRandomForEach([0.5]);
    it("should expand session config to tasks", (): void => {
      const result = providerFixture.expandTasks();
      expect(result).toHaveLength(3);
      expect(result[0]).toBeInstanceOf(Fetch);
    });
  });

  describe("createFetchTestResult", (): void => {
    let responseFixture: Response;

    beforeEach((): void => {
      responseFixture = new Response("", {
        headers: {
          "X-Datacenter": "LCY",
        },
      });
      fetchMock.resetMocks();
      fetchMock.mockResponseOnce(JSON.stringify(clientInfoFixture), {
        headers: {
          "access-control-allow-origin": "*",
        },
      });
    });

    it("should fetch client info", (): Promise<void> => {
      return providerFixture
        .createFetchTestResult(entryFixture, new Response(), taskFixture, {})
        .then((result) => {
          const data = result.beaconData?.data as any;
          expect(fetchMock.mock.calls).toHaveLength(1);
          expect(fetchMock.mock.calls[0][0]).toEqual(
            "https://42c91a26-c33f-482a-9ac9-353cd615c0a9.eu.u.test.fastly-insights.com/l"
          );
          expect(data).toHaveProperty("client_asn");
        });
    });

    it("should get the subject ID from the X-Datacenter response header", (): Promise<
      void
    > => {
      return providerFixture
        .createFetchTestResult(entryFixture, responseFixture, taskFixture, {})
        .then((result) => {
          const clientData = JSON.parse(
            (result.beaconData?.data as any).task_client_data
          );
          expect(clientData.subject_id).toEqual("LCY");
        });
    });

    it("should normalize the resource timing entry", (): Promise<void> => {
      return providerFixture
        .createFetchTestResult(entryFixture, responseFixture, taskFixture, {})
        .then((result) => {
          const clientData = JSON.parse(
            (result.beaconData?.data as any).task_client_data
          );
          expect(clientData).toHaveProperty("subject_request_start");
          expect(clientData).not.toHaveProperty("subject_name");
        });
    });

    it("should stringify the task client data", (): Promise<void> => {
      return providerFixture
        .createFetchTestResult(entryFixture, responseFixture, taskFixture, {})
        .then((result) => {
          const clientData = (result.beaconData?.data as any).task_client_data;
          expect(clientData).toContain("subject_id");
        });
    });
  });

  describe("getResourceRequestHeaders", (): void => {
    // Currently a no-op
  });

  describe("getResourceUrl", (): void => {
    it("should memoize the result for subsquent calls", () => {
      const templateSpy = jest.spyOn(template, "templateResource");

      // Call 3 times
      providerFixture.getResourceUrl(taskFixture);
      providerFixture.getResourceUrl(taskFixture);
      providerFixture.getResourceUrl(taskFixture);

      expect(templateSpy).toHaveBeenCalledTimes(1);
    });

    it("should template the resource", () => {
      const fixture = Object.assign({}, taskFixture);
      fixture.resource =
        "https://<%TEST_ID%>.fastly-insights.com/o.svg?id=<%TEST_ID%>";
      const result = providerFixture.getResourceUrl(fixture);
      expect(result).toEqual(
        "https://42c91a26-c33f-482a-9ac9-353cd615c0a9.fastly-insights.com/o.svg?id=42c91a26-c33f-482a-9ac9-353cd615c0a9"
      );
    });
  });

  describe("handleError", (): void => {
    // Currently a no-op
  });

  describe("makeBeaconData", (): void => {
    let testDataFixture: TestResultBundle;

    beforeAll(() => {
      const FIXED_SYSTEM_TIME = "2021-01-01T00:00:00Z";
      jest.useFakeTimers("modern");
      jest.setSystemTime(Date.parse(FIXED_SYSTEM_TIME));
      return providerFixture
        .createFetchTestResult(entryFixture, new Response(), taskFixture, {})
        .then((result) => {
          testDataFixture = result;
        });
    });

    afterAll(() => {
      jest.useRealTimers();
    });

    it("should add the test_id", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.test_id).toEqual(configFixture.test.id);
    });

    it("should add the test_id", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.test_id).toEqual(configFixture.test.id);
    });

    it("should add the test_api_key", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.test_api_key).toEqual(configFixture.settings.token);
    });

    it("should add the test_lib_version", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.test_lib_version).toEqual(
        configFixture.settings.library_version
      );
    });

    it("should add the server information as a string", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.test_server).toContain(`{"datacenter":"LCY"}`);
    });

    it("should add a unix timestamp", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.test_timestamp).toEqual(1609459200);
    });

    it("should add the task type", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.task_type).toEqual("pop");
    });

    it("should add the task ID", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.task_id).toEqual("LCY");
    });

    it("should add task schema version", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.task_schema_version).toEqual("0.0.0");
    });

    it("should add task server data placeholder", () => {
      const result: any = providerFixture.makeBeaconData(
        taskFixture,
        testDataFixture
      ).data;
      expect(result.task_server_data).toEqual("<% SERVER_DATA %>");
    });
  });

  describe("encodeBeaconData", (): void => {
    beforeAll(() => {
      const FIXED_SYSTEM_TIME = "2021-01-01T00:00:00Z";
      jest.useFakeTimers("modern");
      jest.setSystemTime(Date.parse(FIXED_SYSTEM_TIME));
    });

    afterAll(() => {
      jest.useRealTimers();
    });

    it("should encode the beacon data as a JSON string", () => {
      return providerFixture
        .createFetchTestResult(entryFixture, new Response(), taskFixture, {})
        .then((testDataFixture) => {
          const beaconDataFixture = providerFixture.makeBeaconData(
            taskFixture,
            testDataFixture
          );
          const result = providerFixture.encodeBeaconData(
            taskFixture,
            beaconDataFixture
          );
          expect(result).toMatchSnapshot();
        });
    });
  });

  describe("makeBeaconURL", (): void => {
    it("should construct the beacon URL", () => {
      const result = providerFixture.makeBeaconURL();
      expect(result).toEqual(
        "https://test.fastly-insights.com/b?k=d00fe9b6-91c6-4434-8e77-14630e263a26&s=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJkMDBmZTliNi05MWM2LTQ0MzQtOGU3Ny0xNDYzMGUyNjNhMjYiLCJleHAiOjE1NTY3MDUxNjIsImlhdCI6MTU1NjcwNTEwM30.Pig3FCY94l2vIfBsIHAPsCzE2mgkGpXcbe0QKHPJcq4"
      );
    });
  });
});
