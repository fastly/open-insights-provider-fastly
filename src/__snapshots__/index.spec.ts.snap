// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FastlyProvider encodeBeaconData should encode the beacon data as a JSON string 1`] = `"{\\"test_id\\":\\"42c91a26-c33f-482a-9ac9-353cd615c0a9\\",\\"test_api_key\\":\\"d00fe9b6-91c6-4434-8e77-14630e263a26\\",\\"test_lib_version\\":\\"3.0.0\\",\\"test_server\\":\\"{\\\\\\"datacenter\\\\\\":\\\\\\"LCY\\\\\\"}\\",\\"test_timestamp\\":1609459200,\\"task_type\\":\\"pop\\",\\"task_id\\":\\"LCY\\",\\"task_schema_version\\":\\"0.0.0\\",\\"task_server_data\\":\\"<% SERVER_DATA %>\\",\\"task_client_data\\":\\"{\\\\\\"client_connection\\\\\\":{},\\\\\\"subject_id\\\\\\":\\\\\\"\\\\\\",\\\\\\"subject_attempted_id\\\\\\":\\\\\\"LCY\\\\\\",\\\\\\"subject_start_time\\\\\\":401.78500000183703,\\\\\\"subject_duration\\\\\\":33.96500000235392,\\\\\\"subject_next_hop_protocol\\\\\\":\\\\\\"http/1.1\\\\\\",\\\\\\"subject_worker_start\\\\\\":0,\\\\\\"subject_redirect_start\\\\\\":0,\\\\\\"subject_redirect_end\\\\\\":0,\\\\\\"subject_fetch_start\\\\\\":401.78500000183703,\\\\\\"subject_domain_lookup_start\\\\\\":404.2600000029779,\\\\\\"subject_domain_lookup_end\\\\\\":412.0850000035716,\\\\\\"subject_connect_start\\\\\\":412.0850000035716,\\\\\\"subject_connect_end\\\\\\":428.50000000180444,\\\\\\"subject_secure_connection_start\\\\\\":417.48000000370666,\\\\\\"subject_request_start\\\\\\":428.57000000367407,\\\\\\"subject_response_start\\\\\\":435.3950000004261,\\\\\\"subject_response_end\\\\\\":435.75000000419095,\\\\\\"subject_transfer_size\\\\\\":711,\\\\\\"subject_encoded_body_size\\\\\\":231,\\\\\\"subject_decoded_body_size\\\\\\":231}\\",\\"client_user_agent\\":\\"abc123\\",\\"client_ip\\":\\"1.2.3.4\\",\\"client_asn\\":10225,\\"client_region\\":\\"abc123\\",\\"client_country_code\\":\\"abc123\\",\\"client_continent_code\\":\\"abc123\\",\\"client_metro_code\\":\\"abc123\\",\\"client_postal_code\\":\\"abc123\\",\\"client_conn_speed\\":\\"abc123\\",\\"client_gmt_offset\\":\\"abc123\\",\\"client_latitude\\":\\"abc123\\",\\"client_longitude\\":\\"abc123\\",\\"resolver_ip\\":\\"abc123\\",\\"resolver_asn\\":33,\\"resolver_region\\":\\"abc123\\",\\"resolver_country_code\\":\\"abc123\\",\\"resolver_continent_code\\":\\"abc123\\",\\"resolver_conn_speed\\":\\"abc123\\",\\"resolver_latitude\\":\\"abc123\\",\\"resolver_longitude\\":\\"abc123\\"}"`;
